bond_app/bond.py

```diff
@@ -39,83 +37,70 @@ def build_authz_url(self, scopes, redirect_uri, state=None):
         was issued to by the OAuth provider.
         :param authz_code: Authorization code from OAuth provider
         :param redirect_uri: redirect url that was used when generating the code - will use default if None
-        :param user_info: Information of the user who issued the request to Bond (not necessarily the same as
-        the username for whom the refresh token was issued by the OAuth provider)
+        :param sam_user_id: Id stored in Sam for user who initiated request
         :return: Two values: datetime when token was issued, username for whom the token was issued
         """
         token_response = self.oauth_adapter.exchange_authz_code(authz_code, redirect_uri)
         jwt_token = JwtToken(token_response.get(FenceKeys.ID_TOKEN), self.user_name_path_expr)
-        user_id = self._fetch_user_id(user_info)
         if FenceKeys.REFRESH_TOKEN not in token_response:
             logging.info(
                 "Exchange authz code did not include refresh token in response.\n{}".format(str(token_response)))
             raise exceptions.BadRequest("authorization response did not include " + FenceKeys.REFRESH_TOKEN)
 
-        if self.refresh_token_store.lookup(user_id, self.provider_name) is not None:
+        if self.refresh_token_store.lookup(sam_user_id, self.provider_name) is not None:
             # clear out any existing information if user has previously linked this account before relinking
-            self.unlink_account(user_info)
-        self.refresh_token_store.save(user_id, token_response.get(FenceKeys.REFRESH_TOKEN), jwt_token.issued_at,
+            self.unlink_account(sam_user_id)
+        self.refresh_token_store.save(sam_user_id, token_response.get(FenceKeys.REFRESH_TOKEN), jwt_token.issued_at,
                                       jwt_token.username, self.provider_name)
         return jwt_token.issued_at, jwt_token.username
 
-    def generate_access_token(self, user_info):
+    def generate_access_token(self, sam_user_id):
         """
         Given a user, lookup their refresh token and use it to generate a new refresh token from their OAuth
-        provider.  If a refresh token cannot be found for the user_id provided, a NotFound will be raised.
-        :param user_info: Information of the user who issued the request to Bond (not necessarily the same as
-        the username for whom the refresh token was issued by the OAuth provider)
+        provider.  If a refresh token cannot be found for the sam_user_id provided, a NotFound will be raised.
+        :param sam_user_id: Id stored in Sam for user who initiated request
         :return: Two values: An Access Token string, datetime when that token expires
         """
-        user_id = self._fetch_user_id(user_info)
-        refresh_token = self.refresh_token_store.lookup(user_id, self.provider_name)
+        refresh_token = self.refresh_token_store.lookup(sam_user_id, self.provider_name)
```

Comment:
Wherever we lookup the Refresh Token for users, we might have a problem after we launch this.  Because I think pretty much everything calls Sam using a pet, which means the refresh tokens currently in the system for linked users are under the pet's key.  Now we're switching to the User's key, which is the right thing to do and what we want, but I think it means that all existing links won't work anymore because we won't be able to find the refresh token for existing links.

Reply:
`self._fetch_user_id(user_info)` connects the pet to the user, so we're storing refresh tokens under the user. We do cache SA keys for the pet instead of the user in the FenceTokenVendingMachine, so we might need to clear those out, but they'll also expire eventually so it is a problem that would go away on its own

Reply:
OK, cool.  Just wanted to double check whether we'd need to alert users as part of this release.  Sounds like we're OK.